execution/evm/execution.go

```diff
@@ -464,6 +509,282 @@ func (c *EngineClient) SetFinal(ctx context.Context, blockHeight uint64) error {
+// The cache is bounded to prevent unbounded memory growth - old entries are pruned.
+func (c *EngineClient) cacheBlockHash(height uint64, hash common.Hash) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	c.blockHashCache[height] = hash
+
+	// Prune old entries to keep cache bounded (keep last ~10 blocks)
+	// This is sufficient since SafeBlockLag is only 2
+	const maxCacheSize = 10
+	if len(c.blockHashCache) > maxCacheSize {
+		var minHeight uint64
+		if height >= maxCacheSize-1 {
+			minHeight = height - (maxCacheSize - 1)
+		}
+		for h := range c.blockHashCache {
+			if h < minHeight {
+				delete(c.blockHashCache, h)
+			}
+		}
+	}
+}
+
+// SetFinalized explicitly sets the finalized block hash.
+// This allows the derivation layer to advance finalization independently.
+// Finalized indicates a block that will never be reorged (e.g., included in DA with sufficient confirmations).
+func (c *EngineClient) SetFinalized(ctx context.Context, blockHash common.Hash) error {
+	c.mu.Lock()
+	c.currentFinalizedBlockHash = blockHash
+	// Finalized implies safe.
+	c.currentSafeBlockHash = blockHash
+	args := engine.ForkchoiceStateV1{
+		HeadBlockHash:      c.currentHeadBlockHash,
+		SafeBlockHash:      c.currentSafeBlockHash,
+		FinalizedBlockHash: c.currentFinalizedBlockHash,
+	}
+	c.mu.Unlock()
+
+	return c.doForkchoiceUpdate(ctx, args, "SetFinalized")
+}
+
+// ResumePayload resumes an in-progress payload build using a stored payloadID.
+// This is used for crash recovery when we have a payloadID but haven't yet
+// retrieved and submitted the payload to the EL.
+//
+// Returns the state root from the payload, or an error if resumption fails.
+// Implements the execution.PayloadResumer interface.
+func (c *EngineClient) ResumePayload(ctx context.Context, payloadIDBytes []byte) (stateRoot []byte, err error) {
+
+	// Convert bytes to PayloadID
+	if len(payloadIDBytes) != 8 {
+		return nil, fmt.Errorf("ResumePayload: invalid payloadID length %d, expected 8", len(payloadIDBytes))
+	}
+	var payloadID engine.PayloadID
+	copy(payloadID[:], payloadIDBytes)
+
+	c.logger.Info().
+		Str("payloadID", payloadID.String()).
+		Msg("ResumePayload: attempting to resume in-progress payload")
+
+	stateRoot, _, err = c.processPayload(ctx, payloadID, nil) // txs = nil for resume
+	return stateRoot, err
+}
+
+// checkIdempotency checks if the block at the given height and timestamp has already been executed.
+// It returns:
+// - stateRoot: non-nil if block is already promoted/finalized (idempotent success)
+// - payloadID: non-nil if block execution was started but not finished (resume needed)
+// - found: true if either of the above is true
+// - err: error during checks
+func (c *EngineClient) checkIdempotency(ctx context.Context, height uint64, timestamp time.Time, txs [][]byte) (stateRoot []byte, payloadID *engine.PayloadID, found bool, err error) {
+	// 1. Check ExecMeta (if store is configured)
+	if c.store != nil {
```

Comment:
TXs are unused. 
The function checks if a block exists at a certain height with a specific timestamp. However, it does not verify if the transactions (txs) match what was previously executed or what exists on the Execution Layer (EL).

Risk: If the sequencer sends a different set of transactions for the same height/timestamp (e.g., due to a bug or malicious actor), this function will return the stateRoot of the previous execution, leading to state divergence between nodes.